stl/inc/format

```diff
@@ -1568,7 +1568,16 @@ private:
             _Arg_type = _Basic_format_arg_type::_Custom_type;
         }
 
-        _Store_impl<_Erased_type>(_Arg_index, _Arg_type, static_cast<_Erased_type>(_Val));
+        if constexpr (is_volatile_v<_Ty>) {
+            // _Erased_type is not a volatile type, so static_cast<_Erased_type> would cast away volatile
+            // _Store_impl uses memcpy for storing the value. We cannot use memcpy with volatile types.
+            // Because of the two reasons we have to make a temporary
+            static_assert(!is_volatile_v<_Erased_type>);
+            const _Erased_type _Temp = _Val;
+            _Store_impl<_Erased_type>(_Arg_index, _Arg_type, _Temp);
+        } else {
+            _Store_impl<_Erased_type>(_Arg_index, _Arg_type, static_cast<_Erased_type>(_Val));
+        }

```

Comment:
While video reviewing this, @barcharcraz and @CaseyCarter helped me understand the history of this code and what the Standard says. After looking at this, we believe that `volatile int` is required to be ill-formed by the Standard, according to https://eel.is/c++draft/format.arg#4 . Specifically,

```cpp
typename Context::template formatter_type<remove_cvref_t<T>>()
  .format(declval<T&>(), declval<Context&>())
```

will pass `declval<volatile int&>()` to `formatter_type<int>`, and that will attempt to bind `.format()`'s parameter `const int&` to the lvalue argument of type `volatile int`, and that is not allowed.

@barcharcraz says that after #2323, the example in #2427 will be accepted (with a technically-undefined memcpying from `volatile`), but the Standard says we should reject this, so the code needs to be changed. I'm not immediately sure what the best change would be, but `static_assert(!is_volatile_v<remove_reference_t<_Ty>>)` (warning: not compiled, I just typed that out) might be a minimal fix. @barcharcraz and @CaseyCarter can help investigate.

Reply:
While it doesn't require library-defined `formatter` specializations to do so, I think the wording may allow a program-defined `formatter` specialization to accept `volatile` arguments. The stripping of volatile in [format.arg]/17.2 suggests to me that the drafters probably just didn't care enough to allow or forbid volatile.